server/core/region_kv.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"context"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+// RegionKV is used to save regions.
+type RegionKV struct {
+	*leveldbKV
+	mu           sync.RWMutex
+	batchRegions map[string]*metapb.Region
+	batchSize    int
+	cacheSize    int
+	flushRate    time.Duration
+	flushTime    time.Time
+	ctx          context.Context
+	cancel       context.CancelFunc
+}
+
+const (
+	//DefaultFlushRegionRate is the ttl to sync the regions to kv storage.
+	defaultFlushRegionRate = 3 * time.Second
+	//DefaultBatchSize is the batch size to save the regions to kv storage.
+	defaultBatchSize = 100
+)
+
+// NewRegionKV returns a kv storage that is used to save regions.
+func NewRegionKV(path string) (*RegionKV, error) {
+	levelDB, err := newLeveldbKV(path)
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	kv := &RegionKV{
+		leveldbKV:    levelDB,
+		batchSize:    defaultBatchSize,
+		flushRate:    defaultFlushRegionRate,
+		batchRegions: make(map[string]*metapb.Region, defaultBatchSize),
+		flushTime:    time.Now().Add(defaultFlushRegionRate),
+		ctx:          ctx,
+		cancel:       cancel,
+	}
+	kv.backgroundFlush()
+	return kv, nil
+}
+
+func (kv *RegionKV) backgroundFlush() {
+	ticker := time.NewTicker(dirtyFlushTick)
+	var (
+		isFlush bool
+		err     error
+	)
+	go func() {
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				kv.mu.RLock()
+				isFlush = kv.flushTime.Before(time.Now())
+				kv.mu.RUnlock()
+				if !isFlush {
+					continue
+				}
+				if err = kv.FlushRegion(); err != nil {
+					log.Error("flush regions error: ", err)
+				}
+			case <-kv.ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+// SaveRegion saves one region to KV.
+func (kv *RegionKV) SaveRegion(region *metapb.Region) error {
+	kv.mu.Lock()
+	defer kv.mu.Unlock()
+	if kv.cacheSize < kv.batchSize {
```

Comment:
kv.cacheSize < kv.batchSize -1

`kv.cacheSize < kv.batchSize` means that the flush will be triggered until the size of `batchRegions` reach to `batchSize+1`